Mathlib/Geometry/Manifold/LocalDiffeomorph.lean

```diff
@@ -0,0 +1,272 @@
+  choose Φ hx heq using hf
+  -- In fact, even `ContMDiffOn I J n f Φ.source`.
+  exact ((Φ.contMDiffOn_toFun).congr heq).contMDiffAt (Φ.open_source.mem_nhds hx)
+
+/-- A local diffeomorphism at `x` is differentiable at `x`. -/
+lemma mdifferentiableAt_of_isLocalDiffeomorphAt (hn : 1 ≤ n) (hf : IsLocalDiffeomorphAt I J n f x) :
+    MDifferentiableAt I J f x :=
+  (contMDiffAt_of_isLocalDiffeomorphAt I J hf).mdifferentiableAt hn
+
+/-- A `C^n` local diffeomorphism on `s` is `C^n` on `s`. -/
+lemma contMDiffOn_of_isLocalDiffeomorphOn (hf : IsLocalDiffeomorphOn I J n f s) :
+    ContMDiffOn I J n f s :=
+  fun x hx ↦ (contMDiffAt_of_isLocalDiffeomorphAt I J (hf ⟨x, hx⟩)).contMDiffWithinAt
+
+/-- A local diffeomorphism on `s` is differentiable on `s`. -/
+lemma mdifferentiableOn_of_isLocalDiffeomorphOn (hn : 1 ≤ n) (hf : IsLocalDiffeomorphOn I J n f s) :
+    MDifferentiableOn I J f s :=
+  (contMDiffOn_of_isLocalDiffeomorphOn I J hf).mdifferentiableOn hn
+
+/-- A `C^n` local diffeomorphism is `C^n`. -/
+lemma contMDiff_of_isLocalDiffeomorph (hf : IsLocalDiffeomorph I J n f) : ContMDiff I J n f :=
+  fun x ↦ contMDiffAt_of_isLocalDiffeomorphAt I J (hf x)
+
+/-- A `C^n` local diffeomorphism is differentiable. -/
+lemma mdifferentiable_of_isLocalDiffeomorph (hn : 1 ≤ n) (hf : IsLocalDiffeomorph I J n f) :
+    MDifferentiable I J f :=
+  fun x ↦ mdifferentiableAt_of_isLocalDiffeomorphAt I J hn (hf x)
+
+/-- A diffeomorphism is a partial diffeomorphism. -/
+def Diffeomorph.toPartialDiffeomorph (h : Diffeomorph I J M N n) :
+    PartialDiffeomorph I J M N n where
+  toPartialEquiv := h.toHomeomorph.toPartialEquiv
+  open_source := isOpen_univ
+  open_target := isOpen_univ
+  contMDiffOn_toFun := fun x _ ↦ h.contMDiff_toFun x
+  contMDiffOn_invFun := fun _ _ ↦ h.symm.contMDiffWithinAt
+
+/-- A `C^n` diffeomorphism is a local diffeomorphism. -/
+lemma Diffeomorph.isLocalDiffeomorph (Φ : M ≃ₘ^n⟮I, J⟯ N) : IsLocalDiffeomorph I J n Φ :=
+  fun _x ↦ ⟨Φ.toPartialDiffeomorph, by trivial, eqOn_refl Φ _⟩
+
+-- FUTURE: if useful, also add "a `PartialDiffeomorph` is a local diffeomorphism on its source"
+
+variable {I J}
+
+/-- A local diffeomorphism on `s` is a local homeomorphism on `s`. -/
+theorem LocalDiffeomorphOn.isLocalHomeomorphOn {s : Set M} (hf : IsLocalDiffeomorphOn I J n f s) :
+    IsLocalHomeomorphOn f s := by
+  apply IsLocalHomeomorphOn.mk
+  intro x hx
+  choose U hyp using hf ⟨x, hx⟩
+  exact ⟨U.toPartialHomeomorph, hyp⟩
+
+/-- A local diffeomorphism is a local homeomorphism. -/
+theorem LocalDiffeomorph.isLocalHomeomorph (hf : IsLocalDiffeomorph I J n f) :
+    IsLocalHomeomorph f := by
+  rw [isLocalHomeomorph_iff_isLocalHomeomorphOn_univ]
+  rw [isLocalDiffeomorph_iff_isLocalDiffeomorphOn_univ] at hf
+  exact LocalDiffeomorphOn.isLocalHomeomorphOn hf
+
+/-- A local diffeomorphism is an open map. -/
+lemma LocalDiffeomorph.isOpenMap (hf : IsLocalDiffeomorph I J n f) : IsOpenMap f :=
+  (LocalDiffeomorph.isLocalHomeomorph hf).isOpenMap
```

Comment:
once you've done the changes to enable dot notation, this proof becomes for instance
```suggestion
  hf.isLocalHomeomorph.isOpenMap
```
Same thing below.